python/packages/autogen-agentchat/src/autogen_agentchat/message_store/_memory_message_store.py

```diff
@@ -0,0 +1,85 @@
+import asyncio
+import time
+from collections import deque
+from typing import Optional, Sequence
+
+from autogen_core._component_config import Component
+from pydantic import BaseModel, Field
+from typing_extensions import Self
+
+from ..messages import BaseAgentEvent, BaseChatMessage, MessageFactory, TextMessage
+from ._message_store import MessageStore
+
+
+class MemoryMessage(BaseModel):
+    """A message stored in memory.
+
+    Args:
+        message (BaseAgentEvent | BaseChatMessage): The message to store.
+        timestamp (int): The timestamp of the message in seconds since epoch.
+    """
+
+    message: BaseAgentEvent | BaseChatMessage | TextMessage
+    ts: int = Field(default_factory=lambda: int(time.time()))
+
+
+class MemoryMessageStoreConfig(BaseModel):
+    ttl_sec: Optional[int] = None
+
+
+class MemoryMessageStore(MessageStore, Component[MemoryMessageStoreConfig]):
+    """A message store that stores messages in memory with optional time-to-live.
+
+    Args:
+        ttl_sec (Optional[int]): Time-to-live in seconds for messages. If None, messages don't expire.
+    """
+
+    component_config_schema = MemoryMessageStoreConfig
+    component_provider_override = "autogen_agentchat.message_store.MemoryMessageStore"
+
+    def __init__(self, message_factory: MessageFactory, ttl_sec: Optional[int] = None):
```

Comment:
I think there needs to be a way to "pin" some messages in the store. Otherwise, the user's task message which is often the first one, and the most recent messages, which are necessary for orchestration logics such as Swarm, they shouldn't be allowed to expire because it will break orchestration.

Reply:
For example in Swarm, we need to pin the most recent `HandoffMessage`. If that message is not pinned, it will introduce a runtime error next time the chat manager selects the next agent if the message is expired.

Reply:
Good point. Let me think how to pin some messages. Thanks for your detail feedback.

Reply:
@ekzhu 
I think GroupChatManager pinning or unpinning certain messages is beyond the scope of this PR. How does GroupChatManager know which messages to pin/unpin?
So my conclusion is that the "pin" function is only used inside the ListMessageStore class. What do you think of below approach?
```python
class ListMessageStore(MessageStore, Component[ListMessageStoreConfig]):
    def __init__(self,
                 message_factory: MessageFactory,
                 pin_first_message: bool = False,  # <== added
                 pin_last_message: bool = False,  # <== added
                 ttl_sec: Optional[int] = None):
    ...
```
We can ignore first or last messages in _remove_expired_messages function properly.

Reply:
@ekzhu 
Please refer https://github.com/microsoft/autogen/pull/6350/commits/c75157e405731d5fd3fcf76544a95315ed1e986b, but this is under open discussion now.